esp8266/timer.c

```diff
@@ -0,0 +1,139 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2013, 2014 Damien P. George
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <stddef.h>
+
+#include "py/nlr.h"
+#include "py/runtime.h"
+#include "py/gc.h"
+
+
+#include "osapi.h"
+#include "os_type.h"
+
+#include "gccollect.h"
+
+#include "timer.h"
+#include "utils.h"
+
+extern void ets_timer_disarm(ETSTimer *ptimer);
+extern void ets_timer_setfn(ETSTimer *ptimer, ETSTimerFunc *pfunction, void *parg);
+extern void ets_timer_arm_new(ETSTimer *ptimer,uint32_t milliseconds, bool repeat_flag, bool);
+
+typedef struct _esp_timer_obj_t {
+    mp_obj_base_t base;
+    os_timer_t timer;
+    mp_obj_t callback;
+    uint32_t period;    // in milliseconds
+    bool repeat;
+    mp_obj_t arg;
+    // start_time; // to calculate elapsed, time to fire etc
+} esp_timer_obj_t;
+
+
+STATIC void esp_timer_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    esp_timer_obj_t *self = self_in;
+
+    mp_printf(print, "timer(period=%u, callback=%x, repeat=%c)", 
+        self->period,
+        (unsigned int)&self->callback,
+        self->repeat ? 'y' : 'n');
+}
+
+STATIC void timer_common_callback(void *parg)
+{
+    esp_timer_obj_t *self = (esp_timer_obj_t *)parg;
+    if (self->callback) 
+        call_function_1_protected(self->callback, self->arg);
+    gc_collect();       // done out timed job and are going back to the operating command
+}
+
+STATIC const mp_arg_t esp_timer_init_args[] = {
+    { MP_QSTR_arg,         MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+    { MP_QSTR_repeat,      MP_ARG_KW_ONLY | MP_ARG_INT, {.u_bool = true} }
+};
+#define PYB_TIMER_INIT_NUM_ARGS MP_ARRAY_SIZE(esp_timer_init_args)
+
+STATIC mp_obj_t esp_timer_init_helper(esp_timer_obj_t *self, uint n_args, const mp_obj_t *args, mp_map_t *kw_args) {
```

Comment:
I just copied this verbatim from the other code that handles kwargs. I kind of like the way it makes it quite clear but it's a hack to initialise the vars twice just in case as well. I'll change it.